### ISSUE TITLE
16-video-driver can't pass string

### ISSUE THREAD

**Issue by thederekkaplan** [cfenollosa/os-tutorial#31]
When I pass a string to kprint or kprint_at it always points to null. I have to first declare the string as an array, and then pass it. Why is this happening? I copied/pasted the code.
```
char msg[] = "test";
kprint_at(msg, 1, 2);
```
the above works but the below does not
```
kprint_at("test", 1, 2);
```

**Comment by Vortex1134** [comment]
I'm having a problem where no matter what string i put in the function, i get one single asterisks where the string should start. Also when I use clear_screen() I have a green box on top about 5 or 6 spaces in. Did you ever have that problem? Any ideas on how to solve it?

**Comment by Vortex1134** [comment]
I fixed my one problem but now I have the same issue that you do. Have you ever got it fixed?

**Comment by TravorLZH** [comment]
Check the value of `dh` that indicates the number of sectors to be loaded to the memory. The bytes loaded are equal to number of sectors times 512 bytes/sector. You'd better check the size of kernel image and change that value properly.

**Comment by Menotdan** [comment]
That's probably how to fix it because its probably glitching because not enough disk was read

**Comment by caioaletroca** [comment]
Guys, I have the same problem on a Windows 10, does anybody knows why this happens or have a solution?
Amazing tutorial anyway!!

**Comment by TravorLZH** [comment]
Dear Caio,

I guess this problem is not anything OS-specific, and I believe my previous response will help you solve it.

BTW, Happy new year!

Travor Liu

________________________________

**Comment by caioaletroca** [comment]
No, actually the problems doesn't was about the loaded sectors.
It seems the GCC in windows (dunno if it happens in others systems) when using this environment provided by the tutorial, doesn't have ideia where to put global constants.

Using the knowledge acquired with this tutorial I followed the tutorial from os-dev to Grub Multiboot, and with a linker.ld file with .rodata section solved the problem. Link [here](https://wiki.osdev.org/Bare_Bones)

The .rodata section defines global constants (read-only).
Awesome tutorial, and Happy New Year!!! :D

**Comment by jorge-correia** [comment]
Hey guys, I had the same problem, and @caioaletroca is right. I wasn't using a cross-compiler, so my GCC was generating code for x84_64 architecture and elf64 format. So, our new operating system didn't know where was the literal string.
Now, I am using a cross-compiler (i686-elf-gcc, i686-elf-ld) to generate code in elf32 format and i386 architecture, and it's working.
I recommend this link: [https://wiki.osdev.org/Why_do_I_need_a_Cross_Compiler](Why I need a cross-compiler?)

**Comment by purplewall1206** [comment]
> Check the value of `dh` that indicates the number of sectors to be loaded to the memory. The bytes loaded are equal to number of sectors times 512 bytes/sector. You'd better check the size of kernel image and change that value properly.

which is the right answer , the size of   'os-image.bin' in my device is 5144bytes, change $dh in bootsect.asm from 2 to 13 will fix the problem

**Comment by NOcturnais** [comment]
> I fixed my one problem but now I have the same issue that you do. Have you ever got it fixed?

how did u fix the problem pls ?
